src/test/java/com/fasterxml/jackson/databind/util/concurrentlinkedhashmap/ConcurrentLinkedHashMapTest.java

```diff
@@ -0,0 +1,72 @@
+/*****************************************************************
+ *   Licensed to the Apache Software Foundation (ASF) under one
+ *  or more contributor license agreements.  See the NOTICE file
+ *  distributed with this work for additional information
+ *  regarding copyright ownership.  The ASF licenses this file
+ *  to you under the Apache License, Version 2.0 (the
+ *  "License"); you may not use this file except in compliance
+ *  with the License.  You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ ****************************************************************/
+package com.fasterxml.jackson.databind.util.concurrentlinkedhashmap;
+
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNull;
+
+//copied from https://github.com/apache/cayenne/blob/b156addac1c8e4079fa88e977fee609210c5da69/cayenne-server/src/test/java/org/apache/cayenne/util/concurrentlinkedhashmap/ConcurrentLinkedHashMapTest.java
+public class ConcurrentLinkedHashMapTest {
```

Comment:
I'm not sure which commit Cayenne forked from, but it does look older than my last release. That should be okay, with the caveat that I worked on this library from 2008-2013 so could be a little rusty.

A really nice, simple sanity test is to use [Guava's testlib](https://github.com/google/guava/tree/master/guava-testlib) ([example](https://github.com/ben-manes/caffeine/blob/master/guava/src/test/java/com/github/benmanes/caffeine/guava/MapTestFactory.java)). If I recall correctly, there was one tiny mistake by supporting `entrySet().add(entry)`, whereas Map's contract disallows that. You can use `suppressing(method)` to skip that one. This will give you good coverage for non-LRU / non-concurrent aspects.

Reply:
Thanks @ben-manes for all the pointers. I'll see what I can do with Guava's testlib. I'll also look into adding stress tests based on lincheck if the Guava testlib doesn't provide enough coverage there.

Reply:
The advantage of the Cayenne version is that their copy has no annotations. I wasn't really sure how to replace all the GuardedBy annotations in v1.4.2 - as in, would I need to wrap all the access to code with a GuardedBy annotation with lock acquires and releases?

Reply:
Oh, those are purely documentary annotations from [Java Concurrency in Practice](https://jcip.net/). They only provide hints for the reader as to expectations, but do not modify the compilation process. Those annotations were only available as a standalone jar, so they were re-packaged by jsr305 (since abandoned) which was made available on Maven Central. Much later static analyzers (CheckerFramework, ErrorProne) repackaged them and added support for detecting obvious misuse. At that time the Java static analysis tools were not very mature so these types of documentation annotations (e.g. `@Nullable`) were a best practice as reminders (since few read docs) and no one expected them to modifying behavior or be verified.

Reply:
Let's get rid of these annotations. While I can see benefit from documentation perspective (and maybe newer analysis tools), this is an internal "shaded" copy of something else, and while `provided` dependencies are ok it seems unnecessary for this usage.

Reply:
Agree. I'd recommend just using `// @GuardedBy` and similar, which would have equivalent semantics.

<details>
<summary>4.5. Documenting Synchronization Policies</summary>

Documentation is one of the most powerful (and, sadly, most underutilized) tools for managing thread safety. Users
look to the documentation to find out if a class is thread‐safe, and maintainers look to the documentation to understand the implementation strategy so they can maintain it without inadvertently compromising safety. Unfortunately, both of these constituencies usually find less information in the documentation than they'd like.

__Document a class's thread safety guarantees for its clients; document its synchronization policy for its maintainers.__

Each use of synchronized, volatile, or any thread‐safe class reflects a synchronization policy defining a strategy for
ensuring the integrity of data in the face of concurrent access. That policy is an element of your program's design, and should be documented. Of course, the best time to document design decisions is at design time. Weeks or months later, the details may be a blur ‐ so write it down before you forget.

Crafting a synchronization policy requires a number of decisions: which variables to make volatile, which variables to
guard with locks, which lock(s) guard which variables, which variables to make immutable or confine to a thread, which operations must be atomic, etc. Some of these are strictly implementation details and should be documented for the sake of future maintainers, but some affect the publicly observable locking behavior of your class and should be documented as part of its specification.

At the very least, document the thread safety guarantees made by a class. Is it thread‐safe? Does it make callbacks with a lock held? Are there any specific locks that affect its behavior? Don't force clients to make risky guesses. If you don't want to commit to supporting client‐side locking, that's fine, but say so. If you want clients to be able to create new atomic operations on your class, as we did in Section 4.4, you need to document which locks they should acquire to do so safely. If you use locks to guard state, document this for future maintainers, because it's so easy  ‐ the `@GuardedBy` annotation will do the trick. If you use more subtle means to maintain thread safety, document them because they may not be obvious to maintainers.

The current state of affairs in thread safety documentation, even in the platform library classes, is not encouraging. How many times have you looked at the Javadoc for a class and wondered whether it was thread‐safe? Most classes don't offer any clue either way. Many official Java technology specifications, such as servlets and JDBC, woefully
under-document their thread safety promises and requirements.

While prudence suggests that we not assume behaviors that aren't part of the specification, we have work to get done, and we are often faced with a choice of bad assumptions. Should we assume an object is thread‐safe because it seems that it ought to be? Should we assume that access to an object can be made thread‐safe by acquiring its lock first? (This risky technique works only if we control all the code that accesses that object; otherwise, it provides only the illusion of thread safety.) Neither choice is very satisfying.

To make matters worse, our intuition may often be wrong on which classes are "probably thread‐safe" and which are
not. As an example, java.text.SimpleDateFormat isn't thread‐safe, but the Javadoc neglected to mention this until
JDK 1.4. That this particular class isn't thread‐safe comes as a surprise to many developers. How many programs
mistakenly create a shared instance of a non‐thread‐safe object and used it from multiple threads, unaware that this
might cause erroneous results under heavy load?

The problem with SimpleDateFormat could be avoided by not assuming a class is thread‐safe if it doesn't say so. On the other hand, it is impossible to develop a servlet‐based application without making some pretty questionable
assumptions about the thread safety of container‐provided objects like HttpSession. Don't make your customers or
colleagues have to make guesses like this.

</details>